assignment2/2b/src/Tree.java

```diff
@@ -0,0 +1,122 @@
+                    left.left.parent = left;
+                }
+            } else {
+                right = new Tree<T>(newValue, this, null, right);
+                if (right.right != null) {
+                    right.right.parent = right;
+                }
+            }
+            return true;
+        } else {
+            return (left != null && left.add(newValue, parentVal, isLeft))
+                    || (right != null && right.add(newValue, parentVal, isLeft));
+        }
+    }
+
+    /**
+     *
+     * Get common ancestor
+     *
+     * @param element1 -- first element
+     * @param element2 -- second element
+     * @return -- common ancestors for elements
+     * @throws TreeException.ElementDoesntFound -- if element isn't in tree
+     * @throws TreeException.NoCommonAncestor -- if one of elements is root
+     */
+    T getCommonAncestor(T element1, T element2)
+            throws TreeException.ElementDoesntFound, TreeException.NoCommonAncestor {
+        if (value.equals(element1) || value.equals(element2)) {
+            throw new TreeException.NoCommonAncestor("No common ancestor for values " + element1.toString()
+                    + " and " + element2.toString());
+        }
+
+        List<T> parents1 = getParents(element1);
+        List<T> parents2 = getParents(element2);
+        if (parents1.isEmpty() && !value.equals(element1)) {
+            throw new TreeException.ElementDoesntFound("element " + element1.toString() + " doesn't found");
+        } else if (parents2.isEmpty() && !value.equals(element2)) {
+            throw new TreeException.ElementDoesntFound("element " + element2.toString() + " doesn't found");
+        } else {
+            for (T i : parents1) {
+                if (parents2.contains(i)) {
+                    return i;
+                }
+            }
+        }
+
+        throw new TreeException.NoCommonAncestor("No common ancestor for values " + element1.toString()
+                + " and " + element2.toString());
+    }
+
+    /**
+     * Get list of parents for element. If element doesn't exist, return empty list
+     *
+     * @param element -- value of element
+     * @return -- list of ancestors
+     */
+    private List<T> getParents(T element) {
+        if (value.equals(element)) {
+            return getParents();
+        } else {
+            List<T> leftRes = left == null ? new LinkedList<T>() : left.getParents(element);
+            List<T> rightRes = right == null ? new LinkedList<T>() : right.getParents(element);
+            if (leftRes.isEmpty()) {
+                return rightRes;
+            } else {
+                return leftRes;
+            }
+        }
+    }
+
+    /**
+     * Get parents for current element
+     *
+     * @return list of ancestors
+     */
+    private List<T> getParents() {
+        if (parent == null) {
+            return new LinkedList<T>();
+        } else {
+            List<T> res = parent.getParents();
```

Comment:
Recursion like this isn't very natural in Java, when you might just use a loop instead. When you care about the order, you can add the elements to the List and then reverse it.